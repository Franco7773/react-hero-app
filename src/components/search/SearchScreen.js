import React, { useMemo } from 'react'
import queryString from 'query-string';

import { HeroCard } from '../heroes/HeroCard';
import { useForm } from '../../hooks/useForm';
import { useLocation } from 'react-router-dom';
import { getHeroesByName } from '../../selectors/getHeroesByName';

export const SearchScreen = ({ history }) => {

  const location = useLocation();
  const { q = '' } = queryString.parse( location.search );
  
  const [{ searchText }, handleInputChange ] = useForm({
    searchText: q
  });
  
  const heroesFiltered = useMemo(() => getHeroesByName( q ), [ q ]);

  const handleSubmit = ( event ) => {
    event.preventDefault();

    history.push(`?q=${ searchText }`)
  };
  
  return (
    <div>
      <h1>Search Screen</h1>
      <hr />

      <div className="row">
        <div className="col-4">
          <h4>Search Form</h4>
          <hr />

          <form onSubmit={ handleSubmit }>
            <input type="text" name="searchText" value={ searchText } onChange={ handleInputChange } placeholder="Find your Hero" autoComplete="off" className="form-control"/>

            <button type="submit" className="btn btn-block btn-outline-primary m-1">Search...</button>
          </form>
        </div>

        <div className="col-7">
          <h4>Results</h4>
          <hr />

          {(q === '') && <div className="alert alert-info">Search a Hero</div>}
          {(q !== '' && heroesFiltered.length === 0) && <div className="alert alert-danger">There is no a Hero with { q }</div>}

          {
            heroesFiltered.map( hero => (
              <HeroCard key={ hero.id } { ...hero }/>
            ))
          }
        </div>
      </div>
    </div>
  )
}
