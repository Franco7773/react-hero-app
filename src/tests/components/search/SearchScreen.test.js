import React from 'react';
import { mount } from 'enzyme';
import { MemoryRouter, Route } from 'react-router-dom';
import { SearchScreen } from '../../../components/search/SearchScreen';

describe('Prueba en el <SearchScreen />', () => {
  
  test('Debe de mostrarse correctamente con valores por defecto', () => {
      
    const wrapper = mount(
      <MemoryRouter initialEntries={['/search']}>
        <Route path="/search" component={ SearchScreen } />
      </MemoryRouter>
    );
    
    expect( wrapper ).toMatchSnapshot();
    expect( wrapper.find('.alert-info').text().trim()).toBe('Search a Hero');
  });

  test('Debe de mostrar a batman y el input con el valor del queryString', () => {
    
    const wrapper = mount(
      <MemoryRouter initialEntries={['/search?q=batman']}>
        <Route path="/search" component={ SearchScreen } />
      </MemoryRouter>
    );

    expect( wrapper.find('input').prop('value')).toBe('batman');
    expect( wrapper ).toMatchSnapshot();
  });

  test('Debe de mostrar un error si no se encuentra el hero', () => {
    
    const wrapper = mount(
      <MemoryRouter initialEntries={['/search?q=spiderman']}>
        <Route path="/search" component={ SearchScreen } />
      </MemoryRouter>
    );

    expect( wrapper.find('.alert-danger').text().trim()).toBe('There is no a Hero with spiderman');
  });
  
  test('Debe de llamar el Push del history', () => {
    
    const historyMock = {
      push: jest.fn()
    };

    const wrapper = mount(
      <MemoryRouter initialEntries={['/search?q=spiderman']}>
        <Route path="/search" component={ () => <SearchScreen history={ historyMock } /> } />
      </MemoryRouter>
    );

    wrapper.find('input').simulate('change', {
      target: { name: 'searchText', value: 'batman'}
    });

    wrapper.find('form').prop('onSubmit')({ preventDefault(){} });

    expect( historyMock.push ).toHaveBeenCalledWith('?q=batman');
  });
});
