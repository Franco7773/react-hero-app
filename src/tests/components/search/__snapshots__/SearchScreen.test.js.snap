// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Prueba en el <SearchScreen /> Debe de mostrar a batman y el input con el valor del queryString 1`] = `
<div>
  <h1>
    Search Screen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-4"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Find your Hero"
          type="text"
          value="batman"
        />
        <button
          className="btn btn-block btn-outline-primary m-1"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="card ms-3"
        style={
          Object {
            "masWidth": 540,
          }
        }
      >
        <div
          className="row no-gutters"
        >
          <div
            className="col-md-4"
          >
            <img
              alt="Batman"
              className="card-img"
              src="./assets/heroes/dc-batman.jpg"
            />
          </div>
          <div
            className="col-md-8"
          >
            <div
              className="card-body"
            >
              <h5
                className="card-title"
              >
                Batman
              </h5>
              <p
                className="card-text"
              >
                Bruce Wayne
              </p>
              <p
                className="card-text"
              >
                <small
                  className="text-muted"
                >
                  Detective Comics #27
                </small>
              </p>
              <a
                href="/hero/dc-batman"
                onClick={[Function]}
              >
                MÃ¡s...
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Prueba en el <SearchScreen /> Debe de mostrarse correctamente con valores por defecto 1`] = `
<div>
  <h1>
    Search Screen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-4"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Find your Hero"
          type="text"
          value=""
        />
        <button
          className="btn btn-block btn-outline-primary m-1"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        Search a Hero
      </div>
    </div>
  </div>
</div>
`;
